tree/tree.go

```diff
@@ -0,0 +1,160 @@
+	_, err := t.Get(wTx, key)
+	if err == arbo.ErrKeyNotFound {
+		// key does not exist, use Add
+		return t.Add(wTx, key, value)
+	} else if err != nil {
+		return err
+	}
+	// key already exists, use Update
+	return t.t.UpdateWithTx(wTx, key, value)
+}
+
+// Add adds a new leaf, if the key already exists will return error
+func (t *Tree) Add(wTx db.WriteTx, key, value []byte) error {
+	return t.t.AddWithTx(wTx, key, value)
+}
+
+// AddBatch adds a batch of key-values to the Tree. Returns an array containing
+// the indexes of the keys failed to add. Supports empty values as input
+// parameters, which is equivalent to 0 valued byte array.
+func (t *Tree) AddBatch(wTx db.WriteTx, keys, values [][]byte) ([]int, error) {
+	return t.t.AddBatchWithTx(wTx, keys, values)
+}
+
+func (t *Tree) Iterate(callback func(key, value []byte) bool) error {
+	callbackWrapper := func(currLvl int, k, v []byte) bool {
+		callback(k, v)
+		return true
+	}
+	rTx := t.db.ReadTx()
+	root, err := t.Root(rTx)
+	if err != nil {
+		return err
+	}
+	return t.t.IterateWithStop(root, callbackWrapper)
+}
+
+// Root returns the current root.
+func (t *Tree) Root(rTx db.ReadTx) ([]byte, error) {
+	return t.t.RootWithTx(rTx)
+}
+
+// Size returns the number of leafs under the current root
+func (t *Tree) Size(rTx db.ReadTx) uint64 {
+	// TODO we need to review where this would be used and in which
+	// context, and define the way how this is computed
+	return 0
+}
+
+// GenProof returns a byte array with the necessary data to verify that the
+// key&value are in a leaf under the current root
+func (t *Tree) GenProof(rTx db.ReadTx, key []byte) ([]byte, error) {
+	// TODO NOTE leafK & leafV will be needed for verifying the proof,
+	// maybe is worth to return them
+	_, _, s, existence, err := t.t.GenProofWithTx(rTx, key)
+	if err != nil {
+		return nil, err
+	}
+	if !existence {
+		// proof of non-existence currently not needed in vocdoni-node
+		return nil, fmt.Errorf("key does not exist %s", hex.EncodeToString(key))
+	}
+	return s, nil
+}
+
+// VerifyProof checks the proof for the given key, value and root, using the
+// hash function of the Tree
+func (t *Tree) VerifyProof(key, value, proof, root []byte) (bool, error) {
+	return arbo.CheckProof(t.t.HashFunction(), key, value, root, proof)
+}
+
+// FromRoot returns a new read-only Tree for the given root, that uses the same
+// underlying db.
+func (t *Tree) FromRoot(root []byte) (*Tree, error) {
+	tree, err := t.t.Snapshot(root)
+	if err != nil {
+		return nil, err
+	}
+	return &Tree{
+		t:  tree,
+		db: t.db,
+	}, nil
+}
+
+// Dump exports all the Tree leafs in a byte array.
+func (t *Tree) Dump() ([]byte, error) {
+	// TODO NOTE maybe here we want to allow to specify which root to be
+	// used for the Dump, not just the last one
+	return t.t.Dump(nil)
+}
+
+// ImportDump imports the leafs (that have been exported with the Dump method)
+// in the Tree.
+func (t *Tree) ImportDump(b []byte) error {
```

Comment:
I have a question regarding how this is supposed to be used.  I understand that the use case is that you open a clean tree for the first time, and then you call ImportDump.  Is that right?

What happens if the tree already has leaves and you do ImportDump?  I think this would be very problematic right?  Because you will end up with a different tree than the exported tree.  I've checked the Arbo implementation and it seems there's no special check.  What about checking that the tree has root == emptyHash, and otherwise return error?  And then specify in the function documentation that the tree must be empty before calling ImportDump?

Reply:
Agree, updated